File: src/packages/control-ui/src/api/relay_stream_codec.ts
/**
 * Relay stream network types.
 */
export type RelayNetwork = 'tcp' | 'udp';

/**
 * Relay stream header defining egress tunnel destination.
 * Directly mirrors lumicore::relay::relay_stream_codec.
 */
export interface RelayStreamHeader {
  network: RelayNetwork;
  host: string;
  port: number;
}

/**
 * Encodes into ASCII delimiter format: `<network>@<host>$<port>\r`.
 */
export function encodeRelayV1(header: RelayStreamHeader): Uint8Array {
  const s = `${header.network}@${header.host}$${header.port}\r`;
  return new TextEncoder().encode(s);
}

/**
 * Decodes an ASCII delimiter formatted relay header.
 */
export function decodeRelayV1(src: Uint8Array): {
  header: RelayStreamHeader;
  bytesConsumed: number;
} {
  if (src.length === 0) {
    throw new Error('Empty buffer');
  }

  let crPos = -1;
  for (let i = 0; i < src.length; i++) {
    if (src[i] === 0x0d) {
      // '\r'
      crPos = i;
      break;
    }
  }
  if (crPos === -1) {
    throw new Error('Missing delimiter \\r');
  }

  const text = new TextDecoder().decode(src.subarray(0, crPos));
  const atPos = text.indexOf('@');
  if (atPos === -1) {
    throw new Error('Missing delimiter @');
  }

  const netStr = text.substring(0, atPos).toLowerCase();
  if (netStr !== 'tcp' && netStr !== 'udp') {
    throw new Error(`Invalid relay network: ${netStr}`);
  }

  const rest = text.substring(atPos + 1);
  const dollarPos = rest.indexOf('$');
  if (dollarPos === -1) {
    throw new Error('Missing delimiter $');
  }

  const host = rest.substring(0, dollarPos);
  const portStr = rest.substring(dollarPos + 1);
  const port = parseInt(portStr, 10);
  if (isNaN(port) || port < 1 || port > 65535) {
    throw new Error(`Invalid port: ${portStr}`);
  }

  return {
    header: {
      network: netStr,
      host,
      port,
    },
    bytesConsumed: crPos + 1,
  };
}

/**
 * Encodes into compact binary wire format:
 * Magic [0x52, 0x32] ('R2') + NetByte (1=TCP, 2=UDP) + Port (uint16 BE) + HostLen (uint8) + HostBytes
 */
export function encodeRelayV2(header: RelayStreamHeader): Uint8Array {
  const hostBytes = new TextEncoder().encode(header.host);
  const clampedLen = Math.min(hostBytes.length, 255);
  const out = new Uint8Array(6 + clampedLen);

  out[0] = 0x52; // 'R'
  out[1] = 0x32; // '2'
  out[2] = header.network === 'tcp' ? 1 : 2;
  out[3] = (header.port >> 8) & 0xff;
  out[4] = header.port & 0xff;
  out[5] = clampedLen;
  out.set(hostBytes.subarray(0, clampedLen), 6);

  return out;
}

/**
 * Decodes a compact binary relay header.
 */
export function decodeRelayV2(src: Uint8Array): {
  header: RelayStreamHeader;
  bytesConsumed: number;
} {
  if (src.length < 6) {
    throw new Error('Buffer too short for relay v2 header');
  }

  if (src[0] !== 0x52 || src[1] !== 0x32) {
    throw new Error('Invalid relay v2 magic bytes');
  }

  const netByte = src[2];
  let network: RelayNetwork;
  if (netByte === 1) {
    network = 'tcp';
  } else if (netByte === 2) {
    network = 'udp';
  } else {
    throw new Error(`Unknown network byte: ${netByte}`);
  }

  const port = (src[3]! << 8) | src[4]!;
  const hostLen = src[5]!;
  const totalLen = 6 + hostLen;

  if (src.length < totalLen) {
    throw new Error('Buffer truncated for relay v2 host');
  }

  const host = new TextDecoder().decode(src.subarray(6, totalLen));

  return {
    header: {
      network,
      host,
      port,
    },
    bytesConsumed: totalLen,
  };
}

File: labs/mobile/donor-archive/tunnel/RelayStreamCodec.kt
package com.luminet.android.tunnel

import java.nio.ByteBuffer
import java.nio.charset.StandardCharsets

/**
 * Supported relay forwarding networks.
 */
enum class RelayNetwork(val id: String) {
    TCP("tcp"),
    UDP("udp")
}

/**
 * Relay stream header defining egress tunnel destination.
 * Mirrors lumicore::relay::relay_stream_codec.
 */
data class RelayStreamHeader(
    val network: RelayNetwork,
    val host: String,
    val port: Int
) {
    /**
     * Encodes into ASCII wire format: `<network>@<host>$<port>\r`.
     */
    fun encodeV1(): ByteArray {
        val s = "${network.id}@$host\$$port\r"
        return s.toByteArray(StandardCharsets.UTF_8)
    }

    /**
     * Encodes into compact binary wire format:
     * Magic [0x52, 0x32] ('R2') + NetByte (1=TCP, 2=UDP) + Port (u16 BE) + HostLen (u8) + HostBytes
     */
    fun encodeV2(): ByteArray {
        val hostBytes = host.toByteArray(StandardCharsets.UTF_8)
        val clampedHostLen = hostBytes.size.coerceAtMost(255)
        val bb = ByteBuffer.allocate(6 + clampedHostLen)
        bb.put('R'.code.toByte())
        bb.put('2'.code.toByte())
        bb.put(if (network == RelayNetwork.TCP) 1.toByte() else 2.toByte())
        bb.putShort(port.toShort())
        bb.put(clampedHostLen.toByte())
        bb.put(hostBytes, 0, clampedHostLen)
        return bb.array()
    }

    companion object {
        /**
         * Decodes ASCII delimiter wire format.
         */
        fun decodeV1(src: ByteArray): Pair<RelayStreamHeader, Int> {
            require(src.isNotEmpty()) { "Empty buffer" }
            val crPos = src.indexOf('\r'.code.toByte())
            require(crPos != -1) { "Missing delimiter \\r" }

            val slice = src.copyOfRange(0, crPos)
            val atPos = slice.indexOf('@'.code.toByte())
            require(atPos != -1) { "Missing delimiter @" }

            val netStr = String(slice.copyOfRange(0, atPos), StandardCharsets.UTF_8).lowercase()
            val network = when (netStr) {
                "tcp" -> RelayNetwork.TCP
                "udp" -> RelayNetwork.UDP
                else -> throw IllegalArgumentException("Invalid relay network: $netStr")
            }

            val rest = slice.copyOfRange(atPos + 1, slice.size)
            val dollarPos = rest.indexOf('$'.code.toByte())
            require(dollarPos != -1) { "Missing delimiter $" }

            val host = String(rest.copyOfRange(0, dollarPos), StandardCharsets.UTF_8)
            val portStr = String(rest.copyOfRange(dollarPos + 1, rest.size), StandardCharsets.UTF_8)
            val port = portStr.toInt()

            return Pair(RelayStreamHeader(network, host, port), crPos + 1)
        }

        /**
         * Decodes compact binary wire format.
         */
        fun decodeV2(src: ByteArray): Pair<RelayStreamHeader, Int> {
            require(src.size >= 6) { "Buffer too short for relay v2 header" }
            require(src[0] == 'R'.code.toByte() && src[1] == '2'.code.toByte()) { "Invalid relay v2 magic" }

            val network = when (src[2].toInt()) {
                1 -> RelayNetwork.TCP
                2 -> RelayNetwork.UDP
                else -> throw IllegalArgumentException("Unknown relay network byte: ${src[2]}")
            }

            val bb = ByteBuffer.wrap(src, 3, src.size - 3)
            val port = bb.short.toInt() and 0xFFFF
            val hostLen = bb.get().toInt() and 0xFF
            val totalLen = 6 + hostLen

            require(src.size >= totalLen) { "Buffer truncated for relay v2 host" }
            val hostBytes = ByteArray(hostLen)
            bb.get(hostBytes)
            val host = String(hostBytes, StandardCharsets.UTF_8)

            return Pair(RelayStreamHeader(network, host, port), totalLen)
        }
    }
}
